async_producer.go

```diff
@@ -217,8 +215,15 @@ func (p *asyncProducer) topicDispatcher() {
 		}
 
 		if msg.flags&shutdown != 0 {
-			Logger.Println("Producer shutting down.")
-			break
+			shuttingDown = true
+			continue
+		} else if msg.retries == 0 {
+			if shuttingDown {
+				p.returnError(msg, ErrShuttingDown)
```

Comment:
Oooh, really nice catch. I will fix and see if I can add or adjust a test for this case too.